packages/core/src/api/positionMapping.ts

```diff
@@ -0,0 +1,145 @@
+import type { Transaction } from "prosemirror-state";
+import { Mapping } from "prosemirror-transform";
+import {
+  absolutePositionToRelativePosition,
+  relativePositionToAbsolutePosition,
+  ySyncPluginKey,
+} from "y-prosemirror";
+import type { BlockNoteEditor } from "../editor/BlockNoteEditor.js";
+import * as Y from "yjs";
+import type { ProsemirrorBinding } from "y-prosemirror";
+import type {
+  DefaultInlineContentSchema,
+  DefaultStyleSchema,
+} from "../blocks/defaultBlocks.js";
+import type { DefaultBlockSchema } from "../blocks/defaultBlocks.js";
+import type {
+  BlockSchema,
+  InlineContentSchema,
+  StyleSchema,
+} from "../schema/index.js";
+
+export function isRemoteTransaction(tr: Transaction) {
+  return tr.getMeta(ySyncPluginKey) !== undefined;
+}
+
+type YSyncPluginState = {
+  doc: Y.Doc;
+  binding: Pick<ProsemirrorBinding, "type" | "mapping">;
+};
+
+/**
+ * This is used to keep track of positions of elements in the editor.
+ * It is needed because y-prosemirror's sync plugin can disrupt normal prosemirror position mapping.
+ *
+ * It is specifically made to be able to be used whether the editor is being used in a collaboratively, or single user, providing the same API.
+ */
+export class PositionStorage<
+  BSchema extends BlockSchema = DefaultBlockSchema,
+  ISchema extends InlineContentSchema = DefaultInlineContentSchema,
+  SSchema extends StyleSchema = DefaultStyleSchema
+> {
+  private readonly editor: BlockNoteEditor<BSchema, ISchema, SSchema>;
+  /**
+   * Whether the editor has had a remote transaction.
+   */
+  private hadRemoteTransaction = false;
+
+  private readonly mapping = new Mapping();
+  private mappingLength = 0;
+
+  constructor(
+    editor: BlockNoteEditor<BSchema, ISchema, SSchema>,
+    { shouldMount = true }: { shouldMount?: boolean } = {}
+  ) {
+    this.editor = editor;
+    this.onTransactionHandler = this.onTransactionHandler.bind(this);
+
+    if (!shouldMount) {
+      return;
+    }
+
+    if (!this.editor._tiptapEditor) {
+      throw new Error("Editor not mounted");
+    }
+    this.editor._tiptapEditor.on("transaction", this.onTransactionHandler);
+  }
+
+  /**
+   * This will be called whenever a transaction is applied to the editor.
+   *
+   * It's used to update the position mapping or tell if there was a remote transaction.
+   */
+  private onTransactionHandler({ transaction }: { transaction: Transaction }) {
+    if (this.hadRemoteTransaction) {
+      // If we have already had a remote transaction, we rely only on relative positions, so no need to update the mapping.
+      return;
+    }
+
+    if (isRemoteTransaction(transaction)) {
+      this.hadRemoteTransaction = true;
+    } else {
+      this.mapping.appendMapping(transaction.mapping);
+      this.mappingLength += transaction.mapping.maps.length;
```

Comment:
can't get this from `this.mapping`?

Reply:
ah; after reading further; I suppose the answer is that you can use a single `PositionStorage` object to track multiple positions. That makes sense and is pretty nifty. On the other hand, having a single `PositionStorage` object per tracked position would make it a little bit more efficient I suppose